Rust/src/anomaly_score_visitor.rs

```diff
@@ -0,0 +1,196 @@
+
+    // For improved performance, we set a flag if the point to score lies in
+    // a bounding box. Once this happens, the anomaly score does not update.
+    point_inside_box: bool,
+
+    // Similar to point_inside_box but for each coordinate, allowing
+    // short-cutting of certain computations
+    coordinate_inside_box: Vec<bool>
+}
+
+impl<'a, T> AnomalyScoreVisitor<'a, T>
+where
+    T: Float + One + Sum + Zero
+{
+
+    /// Initialize an anomaly score visitor with a tree and a point to score.
+    ///
+    /// The anomaly score of this visitor is initialized to zero.
+    pub fn new(
+        tree: &'a Tree<T>,
+        point_to_score: &'a Vec<T>,
+    ) -> AnomalyScoreVisitor<'a, T> {
+        AnomalyScoreVisitor {
+            tree: tree,
+            point_to_score: point_to_score,
+            anomaly_score: Zero::zero(),
+            point_inside_box: false,
+            coordinate_inside_box: vec![false; point_to_score.len()]
+        }
+    }
+
+    /// Initialize the anomaly score from a leaf node.
+    ///
+    /// A leaf node is the first node visited in the anomaly scoring process.
+    /// The initial anomaly score depends whether the point at the leaf node
+    /// is equal to the point to score.
+    pub fn accept_leaf(&mut self, leaf: &Leaf, depth: T) {
+        let point_store = self.tree.borrow_point_store();
+        let point = point_store.get(leaf.point()).unwrap();
+        if *self.point_to_score == *point {
+            self.point_inside_box = true;
+            self.anomaly_score = damp::<T>(leaf.mass(), self.tree.mass()) *
+                score_seen(depth, leaf.mass());
+        } else {
+            self.anomaly_score = score_unseen(depth);
+        }
+    }
+
+    /// Update the anomaly score from an internal node.
+    ///
+    /// After visiting a leaf node in [`Self::accept_leaf`] the remaining nodes
+    /// encountered while traversing up the tree to the root are internal
+    /// nodes. The main piece of information used by this algorithm is the
+    /// bounding box at these nodes.
+    pub fn accept(&mut self, node: &Internal<T>, depth: T) {
+        if self.point_inside_box { return; }
+
+        let separation_probability = self.separation_probability(node.bounding_box());
+        if separation_probability <= Zero::zero() {
+            self.point_inside_box = true;
+            return;
+        }
+
+        let one: T = One::one();
+        self.anomaly_score = separation_probability * score_unseen(depth) +
+            (one - separation_probability) * self.anomaly_score;
+    }
+
+    /// Normalize and return the anomaly score computed by the visitor.
+    ///
+    /// The anomaly score computed by the visiting process, via
+    /// [`Self::accept_leaf()`] and [`Self::accept()`], is normalized using the
+    /// mass of the tree before returning. This is so that the resulting anomaly
+    /// score is independent of the number of samples in the tree.
+    pub fn get_result(&self) -> T {
```

Comment:
minor: it seems like we're mostly not using "get" in our getter method names. Should we change this to just `result()`?

Reply:
I guess I saw this `get_result()` not so much as a getter but as a "finalize" like thing. After all, `get_result()` usually does more than just return a value; here it does some normalization. From my reading, I don't see a standard w.r.t. the visitor pattern.

More generally, I chose to omit the `get_`s in the type getters just for brevity. Also, I've seen a number of Rust devs use `field()` and `field_mut()` instead of `get_field()` and `set_field()`. So the question is, do we try to match the Java implementation of RCF or do we write things in the way Rust developers expect.

Anyway, that's my perspective. Given that, willing to adjust in the nomenclature based on your thoughts.

Reply:
My vote is to follow rust conventions when they conflict with what's in the Java library.

Reply:
Right on. In that case I'll leave `get_result()` as is. There is an opportunity to revisit in #206 .